Circadian Archaic Introgression Project
=======================================
` Keila S. Velazquez-Arcelay `


SCRIPTS
-------

Gene and SNP scripts:
- circadian_gene_set.py
  - Imports lists of circadian genes from different sources of evidence. Generates sets of circadian genes at different confidence levels based on the intersection with the sources of evidence.
- circadian_gene_loci.py
  - Creates bed file of circadian genes and their 1Mb flanking loci. Additionally creates a list of circadian genes.
- circadian_variants.py
  - Filter the 1000 Genomes Project variants in circadian genes and flanking within 1Mb.
- circadian_promoter.py
  - Adds the promoter region for each circadian gene. Promoter regions are defined as the region -5kb upstream and 1kb downstream of the gene TSS. TSS sites retrieved from the Biomart.
- circadian_variants_ccres.py
  - Identify candidate cis-regulatory elements (cCREs) in the circadian flanking regions (1Mb). Published in https://doi.org/10.1038/s41586-020-2493-4.
- parse_kuhlwilm19.py
  - Parses the human and archaic variant sites from Kuhlwilm and Boeckx 2019 to find human- and archaic-specific variants.
- circadian_variants_fixed.py
  - Extract circadian variants from a set of human- and archaic-specific variants published by Kuhlwilm and Boeckx 2019.
- circadian_variants_fixed_promoter.py
  - Extract human- and archaic-specific variants in circadian promoter regions. Variant set from Kuhlwilm and Boeckx 2019. Variant set from Kuhlwilm and Boeckx 2019.
- circadian_variants_fixed_ccres.py
  - Human- and archaic-specific circadian variants that are candidate cis-regulatory elements (cCREs). Set of variants from Kuhlwilm and Boeckx 2019.
- circadian_variants_introgressed.py
  - Extracts circadian variants from a set of introgressed variants published by Browning et al., 2018. The set of circadian variants includes variants inside circadian genes, cCREs flanking circadian genes within 1Mb, and promoter variants flanking the TSS by -5kb upstream and 1kb downstream.
- introgression_maps.py
  - Creates a dataset of all the available archaic introgression maps. Study that also merged introgression maps: https://www.science.org/doi/10.1126/sciadv.abc0776
- circadian_genes_divergently_regulated.py
  - Finds divergently regulated circadian genes, defined as PrediXcan predictions with an empirical p-value = 0. General PrediXcan predictions generated by Colbran et al., 2019
- circadian_predixcan_pctl.py
  - Generates percentiles for where the archaic PrediXcan predictions fall in the human distribution. General PrediXcan predictions generated by Colbran et al., 2019
- circadian_variants_splice_altering.py
  - Extract archaic-specific variants in circadian genes, with evidence of having splice-altering effects in at least 1 of the 4 high-coverage archaic genomes. General SAV dataset generated by Colin Brand using SpliceAI predictions.
- circadian_gene_count_by_predixcan_tissue_prediction.py
  - Tests the number of circadian genes contain PrediXcan predictions in each tissue to understand how the difference in power in each GTEx tissue affects the significance values of tissue-specific expression enrichment analysis.
- circadian_gene_count_by_gtex_tissue_rnaseq_expression.py
  - Tests the fraction of circadian genes expressed in each GTEx tissue. Expressed genes are defined as genes containing an RNASeq equal or above 1.



Statistical scripts:
- enrichment_circadian_fixed_snps_genic.py
  - Enrichment analysis of human and archaic lineage-specific SNPs in circadian genes. Set of SNPs from Kuhlwilm and Boeckx, 2019.
- enrichment_circadian_fixed_snps_promoter.py
  - Enrichment analysis of human and archaic lineage-specific SNPs in circadian promoters. Set of SNPs from Kuhlwilm and Boeckx, 2019. 
- enrichment_circadian_fixed_snps_regulatory.py
  - Enrichment analysis of human and archaic lineage-specific SNPs in circadian cCREs with regulatory function. Set of SNPs from Kuhlwilm and Boeckx, 2019.
- enrichment_circadian_introgressed_gtex.py
  - Enrichment analysis of of introgressed variants in circadian genes with evidence of being eQTL in GTEx. Introgressed variants are analyzed in two sets: Variants identified by Browning et al., 2018 or variants identified by Browning and at least one other method.
- enrichment_introgressed_gwas_or_no_gwas.py
  - Test if there is a significant difference between the circadian or not-circadian introgressed variants associated with at least 1 GWAS phenotype or no phenotype.



Other
- opentargets_generate_input.py
  - LiftOver Browning et al., 2018 introgressed variants and generate input file for opentargets.py to scrape associations from a list of variants.
- opentarget.py, opentarget_run.py
  - Scripts written by Sarah Fong. Web-scrape Opentargets genetic associations from a list of SNPs.
- input_for_morningness_cumulative_fraction.py
  - Generate the cumulative fraction of UKBiobank morningness variants sorted by P-value to determine if introgressed variants increase or decreaes morningness. Create input file to plot the cumulative fraction of all introgressed chronotype variants against the set of circadian introgressed chronotype variants.
- table_circadian_genes.py
  - Webscrapes descriptions for each of the candidate circadian genes. Creates Table 1 containing circadian genes, descriptions, and evidence sources.
- table_circadian_genes_sav.py
  - Creates a table containing a list of archaic variants predicted to be splice-altering by SpliceAI. Input: circadian_variants_sav.tsv
- table_circadian_genes_dr.py
  - Creates a table containing a list of divergently regulated genes in each of the archaic hominins. Input: Table 1.
- table_gwas_traits_per_snp.py
  - Generates counts of GWAS phenotypes per introgressed SNP in two sets of SNPs: Circadian and Non-circadian. The GWAS associations were extracted from Opentargets.
- input_for_morningness_cumulative_fraction.py
  - Creates cumulative fraction of p-values associated with the UK Biobank morningness phenotyme in a GWAS analysis.
- input_for_morningness_latitude_cline.py
  - Extracts circadian introgressed variants associated with chronotype in a UKBiobank GWAS. Population latitudes used for linear regression analysis to identify SNPs in this set following a latitudinal cline in 1000 Genomes Projects populations from Eurasian ancestry.



Plots
- plot_upsetr_dr_and_sav.ipynb
  - Plots two UpsetR graphs in R: 1) Number of circadian genes that are divergently regulated (PrediXcan) in a GTEx tissue for each of the 3 archaic hominins, 2) Number of circadian splice-altering variants that are fixed in each of the 4 archaic hominins. 
- plot_predixcan_distribution.ipynb
  - Plot the distribution of gene regulation PrediXcan predictions in 3 archaic hominins and 1000 Genomes Project individuals. PrediXcan values published in Colbran et al., 2019.
- plot_enrichment_introgressed_circadian_in_gtex_tissue.ipynb
  - Plots the probability of success of circadian eQTLs in each GTEx tissue, in a background of all introgressed GTEx eQTLs.
- plot_dr_and_sav_piechart.ipynb
  - Create piechart where the total of values is the number of circadian genes, and the slices represent the number of PrediXcan divergently regulated genes and the number of archaic-specific splice-altering variants.
- plot_introgressed_pleiotropy.ipynb
  - Plots the counts of GWAS phenotypes per introgressed SNP in two sets of SNPs: Circadian and Non-circadian. The GWAS associations were extracted from Opentargets.
- plot_morningness_direction_of_effect.ipynb
  - Plot the cumulative fraction of morningness propensity from a GWAS study on the UKBiobank data (Nealelab).
- plot_morningness_latitude_cline.ipynb
  - Evaluates a set of morningness associated variants that follow a latitudinal cline in allele frequency in Eurasia.
  